File: DSA09014.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

void check_cycle(unordered_map<int, vector<int>> adj, unordered_map<int, bool>& visited, int u, bool& found, int parent) {
    visited[u] = true;
    for (auto x : adj[u]) {
        if (not visited[x]) {
            check_cycle(adj, visited, x, found, u);
        }
        else if (x != parent){
            found = true;
        }
    }
}

int main() {
    int t; cin >> t;
    while (t--) {
        unordered_map<int, vector<int>> adj;
        int v, e; cin >> v >> e;
        for (int i = 0; i < e; i++) {
            int u, v; cin >> u >> v;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        bool found = false;
        unordered_map<int, bool> visited;
        for (auto x : adj) {
            if (not visited[x.first]) {
                check_cycle(adj, visited, x.first, found, -1);
                if (found) break;
            }
        }

        if (found) {
            cout << "YES\n";
        }
        else cout << "NO\n";
    }
}

File: DSA09014.py
from collections import deque, defaultdict


def check_cycle(adj, visited, u, found, parent):
    visited[u] = True
    for x in adj[u]:
        if not visited[x]:
            check_cycle(adj, visited, x, found, u)
        elif x != parent:
            found[0] = True


def function():
    v, e = map(int, input().split())
    ls = list(map(int, input().split()))
    adj = defaultdict(list)
    for i in range(0, 2 * v, 2):
        adj[ls[i]].append(ls[i + 1])
        adj[ls[i + 1]].append(ls[i])

    visited = defaultdict(bool)
    found = [False]
    for x in list(adj):
        if not visited[x]:
            check_cycle(adj, visited, x, found, -1)
            if found[0]:
                break
    print("YES" if found[0] else "NO")


if __name__ == "__main__":
    for i in range(int(input())):
        function()
